### PR TITLE
Introduce complex_imag_exprt and complex_real_exprt

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

File: src/util/std_expr.h [hunk 1]
```diff
@@ -1731,6 +1731,106 @@ inline void validate_expr(const complex_exprt &value)
 }
 
 
+/// \brief Real part
```
Comment (tautschnig) [inline src/util/std_expr.h]: part ... of what? (I'd say "Real part of the expression describing a complex number.")

File: src/util/std_expr.h [hunk 2]
```diff
@@ -1731,6 +1731,106 @@ inline void validate_expr(const complex_exprt &value)
...
+/// \a expr must be known to be a \ref complex_real_exprt.
+///
+/// \param expr: Source expression
+/// \return Object of type \ref complex_real_exprt
+inline const complex_real_exprt &to_complex_real_expr(const exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_complex_real);
+  DATA_INVARIANT(
+    expr.operands().size() == 1,
+    "Real part retrieval operation must have one operand");
```
Comment (tautschnig) [inline src/util/std_expr.h]: Use lowercase for error messages (applies across this PR).

File: src/util/std_expr.h [hunk 3]
```diff
@@ -1731,6 +1731,106 @@ inline void validate_expr(const complex_exprt &value)
...
+{
+  validate_operands(
+    expr,
+    1,
+    "Real part retrieval operation must have one operand");
+}
+
+
+/// \brief Imaginary part
+class complex_imag_exprt:public unary_exprt
```
Comment (tautschnig) [inline src/util/std_expr.h]: I think clang-format wants spaces around `:`

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: c76d96f4d47b21e66bc1324e65093d78d782cd52).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84129475
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: bb6ad6dd3d3dcdd4f9c96d65ba7678d8885b9afe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84132032
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 899fe03c5b722ccb2043f2f60e8756e88be7ffe8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84142660

**Review by chrisr-diffblue — APPROVED** [review]
LGTM - this seems a lot cleaner.

**Pull request merged**